Level 1

letStayFoolish opened the pull request:
Сдача этапа 2: бэкенд

letStayFoolish committed:
First commit

letStayFoolish committed:
Add: controller users, error-handlers, routes-users

letStayFoolish committed:
Add: controller movies.js, errorsHandler.js

letStayFoolish committed:
Add: routes validation using celebrate & Joi

letStayFoolish committed:
Add: sign-in, sign-up, sign-out blank files
Update: Code reformat

letStayFoolish committed:
Update: on the way to find and fix error, I corrected some issues in the code

letStayFoolish committed:
Add: Sign-out

letStayFoolish committed:
Add: Sign-in

letStayFoolish committed:
Add: Sign-out

letStayFoolish committed:
Update: auth.js

letStayFoolish committed:
Update: .then & .catch replaced with async/await

letStayFoolish committed:
Update: movies.js: getFavoriteMovies only for the current user

letStayFoolish committed:
Add: errorLogger & requestLogger (winston, express-winston)

letStayFoolish committed:
Update: errors handling in try {} block.

letStayFoolish committed:
Update: Readme

adzhenalieva commented:
# Резюме:
## Backend

### Критерии, влияющие на работоспособность
- [x] <!--3.64--> В репозитории есть все необходимые инфраструктурные файлы:
  - [x] в проекте есть package.json;
  - [x] .editorconfig;
  - [x] .eslintrc, расширяющий конфигурацию airbnb-base, а также необходимые для работы линтера dev-зависимости;
  - [x] .gitignore.
  - [x] в .eslintrc добавлено исключение для _id. Директивы eslint-disable, eslint-disable-line и eslint-disable-next-line запрещены.
- [ ] <!--3.64--> Нет ошибок линтера. При выполнении команды npx eslint . они должны отсутствовать.
    _https://disk.yandex.ru/i/JbFAvoJ09FOn9g следует исправить ошибки линтера_
- [x] <!--3.64--> В разделе scripts файла package.json есть команды:
  - [x] npm run start, которая запускает сервер на localhost:3000;
  - [x] npm run dev, которая запускает сервер на localhost:3000 с хот релоудом.
- [x] <!--21.81--> Описанные роуты работают корректно:
  - [x] запрос на `GET /users/me` возвращает информацию о пользователе (email и имя);
  - [x] `PATCH /users/me` — обновляет информацию о пользователе;
  - [x] `GET /movies` — все сохранённые пользователем фильмы;
  - [x] `POST /movies` — создаёт фильм с переданными в теле данными;
  - [x] `DELETE /movies/movieId` — удаляет сохранённый фильмы по `_id`;
  - [x] `POST /signup` — создаёт пользователя с переданными в теле данными;
  - [x] `POST /signin` — возвращает JWT, если в теле запроса переданы правильные почта и пароль.
  - [x] Если вы сохраняете JWT в куках, роут `/signout` должен удалять JWT из куки.
- [x] <!--3.64--> Пользователь не может удалить сохранённый фильм из профиля другого пользователя.
- [x] <!--3.64--> Все роуты, кроме /signin и /signup, защищены авторизацией.
- [x] <!--3.64--> Роуты пользователей и роуты фильмов описаны в отдельных файлах.
- [x] <!--7.28--> Ошибки API обрабатываются:
  - [x] используются статус-коды ошибок: 400, 401, 403, 404, 409, 500;
  - [x] если в запросе что-то не так, сервер возвращает статус ответа, который соответствует типу ошибки;
	- [x] сообщение об ошибке соответсвует ее типу;
  - [x] асинхронные обработчики завершены блоком catch;
  - [x] API не возвращает стандартных ошибок базы данных или Node.js.
- [x] <!--3.64--> Реализовано бережное хранение пароля:
  - [x] пароль хранится в виде хеша;
  - [x] API не возвращает хеш пароля клиенту.
- [x] <!--7.28--> Валидация реализована верно:
  - [x] Запрос валидируется перед передачей контроллеру. Тело и, где необходимо, заголовки и параметры проверяются по соответствующим схемам. Если запрос не соответствует схеме, обработка не передаётся контроллеру, и клиент получает ошибку валидации. 
  - [x] Данные валидируются перед добавлением в базу. 
- [x] <!--7.28--> В production-режиме адрес базы данных берётся из process.env.
- [x] <!--3.64--> К серверу можно обратиться по https по адресу домена, указанному в README.md. 
- [x] <!--7.28--> Правильно реализовано хранение секретного ключа для создания JWT:
  - [x] для продакшн сборки он хранится в .env файле — этот файл не добавляется в git;
  - [x] в режиме разработки (при process.env.NODE_ENV !== 'production') код запускается и работает и без .env файла.

### Хорошие практики
- [x] <!--2.14--> Все роуты подключены в файле index.js, который находится в папке routes. Оттуда единый роут подключается в файле app.js.
- [x] <!--2.14--> Асинхронные операции реализованы промисами или async/await.
- [ ] <!--2.14--> Валидация описана в отдельном модуле.
- [x] <!--2.14--> Настроено логгирование:
  - [x] запросы и ответы записываются в файл request.log;
  - [x] ошибки записываются в файл error.log;
  - [x] файлы логов не добавляются в репозиторий.
- [x] <!--2.14--> Ошибки обрабатываются централизованным обработчиком.
- [x] <!--2.14--> Централизованный обработчик ошибок описан в отдельном модуле.
- [x] <!--2.85--> API размещено на отдельном поддомене, например api.movies-explorer.nomoreparties.co или API приложения располагается на домене вида: name.zone/api, а не просто name.zone

### Рекомендации
- [x] <!--0.71--> Для ошибок API созданы классы, расширяющие конструктор Error.
- [x] <!--0.71--> Используется модуль Helmet для установки заголовков, связанных с безопасностью.
- [ ] <!--0.71--> Конфигурация и константы в отдельных файлах:
  - [x] адрес Mongo-сервера в режиме разработки вынесен в отдельный конфигурационный файл;
  - [ ] сообщения ответов и ошибок (константы приложения) вынесены в отдельный файл с константами.
- [x] <!--0.71--> Настроен rate limiter: число запросов с одного IP в единицу времени ограничено.
- [x] <!--0.71--> Rate limiter сконфигурирован в отдельном файле и импортируется в app.js.
- [x] <!--0.71--> Для имени базы данных используется `bitfilmsdb`.

## Количество баллов: 94

## Доп. Комментарии:
- В схему пользователя у поля name следует удалить дефолтное значение, так как это поле обязательное для создания пользователя

letStayFoolish committed:
update: eslit corrections

letStayFoolish committed:
Level-1: Update Signin & Signup

Signin and Signup now return data (user) so we could get user information (name, email) to set currentUser on front-end

letStayFoolish committed:
Level-1: Update: corse: origin

Added domains:
'https://more-movies.nomoredom.nomoredomainsicu.ru/',
 'https://more-movies.nomoredom.nomoredomainsicu.ru/'

 So it could be hosted and tested well

letStayFoolish committed:
Level-1: Update: PORT change to 3007

letStayFoolish committed:
Level-1: Update: PORT change 3007 to 3000

letStayFoolish committed:
Level-1: Update: cors allowance

letStayFoolish committed:
Level-1: Update: cors allowance

letStayFoolish committed:
Level-1: Update: cors allowance

letStayFoolish committed:
Level-1: Update: cors reinstall
